File: nice-traffic-light/r/tests/testthat/test-classifier.R
fixture_path <- function(name) {
  normalizePath(
    file.path("..", "..", "..", "fixtures", paste0(name, ".json")),
    mustWork = TRUE
  )
}

load_fixture <- function(name) {
  jsonlite::fromJSON(fixture_path(name), simplifyVector = FALSE)
}

test_that("near_population parses to known category with no triggers", {
  outcome <- classify_nice_response(load_fixture("near_population"))

  expect_equal(outcome$category, "near_population")
  expect_false(outcome$refer_for_genetics_assessment)
  expect_length(outcome$triggers, 0L)
})

test_that("moderate exposes a single trigger", {
  outcome <- classify_nice_response(load_fixture("moderate"))

  expect_equal(outcome$category, "moderate")
  expect_true(outcome$refer_for_genetics_assessment)
  expect_length(outcome$triggers, 1L)
})

test_that("high exposes all triggers", {
  outcome <- classify_nice_response(load_fixture("high"))

  expect_equal(outcome$category, "high")
  expect_true(outcome$refer_for_genetics_assessment)
  expect_length(outcome$triggers, 2L)
})

test_that("missing cancer_risk block raises response_schema_error", {
  expect_error(
    classify_nice_response(list(model = "NICE")),
    class = "response_schema_error"
  )
})

test_that("unknown category raises response_schema_error", {
  expect_error(
    classify_nice_response(list(
      cancer_risk = list(
        nice_category = "catastrophic",
        nice_refer_genetics = TRUE,
        nice_triggers = list(),
        notes = list()
      )
    )),
    class = "response_schema_error"
  )
})

test_that("non-string trigger raises response_schema_error", {
  expect_error(
    classify_nice_response(list(
      cancer_risk = list(
        nice_category = "moderate",
        nice_refer_genetics = TRUE,
        nice_triggers = list("ok", 42),
        notes = list()
      )
    )),
    class = "response_schema_error"
  )
})

File: nice-traffic-light/python/tests/test_classifier.py
import json
from pathlib import Path

import pytest

from nice_traffic_light.classifier import (
    ResponseSchemaError,
    RiskCategory,
    classify_nice_response,
)

FIXTURES = Path(__file__).resolve().parents[2] / "fixtures"


def _fixture(name: str) -> dict:
    return json.loads((FIXTURES / f"{name}.json").read_text(encoding="utf-8"))


def test_near_population_parses_to_enum_with_no_triggers() -> None:
    outcome = classify_nice_response(_fixture("near_population"))

    assert outcome.category is RiskCategory.NEAR_POPULATION
    assert outcome.refer_for_genetics_assessment is False
    assert outcome.triggers == ()


def test_moderate_exposes_single_trigger() -> None:
    outcome = classify_nice_response(_fixture("moderate"))

    assert outcome.category is RiskCategory.MODERATE
    assert outcome.refer_for_genetics_assessment is True
    assert len(outcome.triggers) == 1


def test_high_exposes_all_triggers_and_refer_flag() -> None:
    outcome = classify_nice_response(_fixture("high"))

    assert outcome.category is RiskCategory.HIGH
    assert outcome.refer_for_genetics_assessment is True
    assert len(outcome.triggers) == 2


def test_missing_cancer_risk_block_raises() -> None:
    with pytest.raises(ResponseSchemaError):
        classify_nice_response({"model": "NICE"})


def test_unknown_category_raises() -> None:
    with pytest.raises(ResponseSchemaError):
        classify_nice_response(
            {
                "cancer_risk": {
                    "nice_category": "catastrophic",
                    "nice_refer_genetics": True,
                    "nice_triggers": [],
                    "notes": [],
                }
            }
        )


def test_non_string_trigger_raises() -> None:
    with pytest.raises(ResponseSchemaError):
        classify_nice_response(
            {
                "cancer_risk": {
                    "nice_category": "moderate",
                    "nice_refer_genetics": True,
                    "nice_triggers": ["ok", 42],
                    "notes": [],
                }
            }
        )
